### PR TITLE
Prepare to make send translog of recovery non-blocking

### PR DESCRIPTION
**Pull request by dnhatn**
This commit prepares the required infra to make send a translog snapshot
of the recovery source non-blocking. I'll make a follow-up to make the send 
snapshot method non-blocking.

Relates #37291

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 1]
```diff
@@ -226,25 +227,28 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
...
-                final long maxSeenAutoIdTimestamp = shard.getMaxSeenAutoIdTimestamp();
-                final long maxSeqNoOfUpdatesOrDeletes = shard.getMaxSeqNoOfUpdatesOrDeletes();
-                sendSnapshotResult = phase2(startingSeqNo, requiredSeqNoRangeStart, endingSeqNo, snapshot,
-                    maxSeenAutoIdTimestamp, maxSeqNoOfUpdatesOrDeletes);
-            } catch (Exception e) {
-                throw new RecoveryEngineException(shard.shardId(), 2, "phase2 failed", e);
-            }
 
+            final Translog.Snapshot phase2Snapshot = shard.getHistoryOperations("peer-recovery", startingSeqNo);
+            resources.add(phase2Snapshot);
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: I'd feel better if we had a finally block around all the code here that closes all resources

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 2]
```diff
@@ -226,25 +227,28 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
...
+            // we can release the retention lock here because the snapshot itself will retain the required operations.
+            IOUtils.close(retentionLock, () -> resources.remove(retentionLock));
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: maybe we just double close?

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 3]
```diff
@@ -538,9 +549,11 @@ SendSnapshotResult phase2(long startingSeqNo, long requiredSeqNoRangeStart, long
         }
 
         final CancellableThreads.IOInterruptible sendBatch = () -> {
-            final long targetCheckpoint = recoveryTarget.indexTranslogOperations(
-                operations, expectedTotalOps, maxSeenAutoIdTimestamp, maxSeqNoOfUpdatesOrDeletes);
-            targetLocalCheckpoint.set(targetCheckpoint);
+            // TODO: Make this non-blocking
+            final PlainActionFuture<Long> future = new PlainActionFuture<>();
+            recoveryTarget.indexTranslogOperations(
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: ++ thanks for keeping it simple.

**Review by s1monw — APPROVED** [review]
LGTM left some suggestions

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java (hunk 2 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: +1. I pushed https://github.com/elastic/elasticsearch/pull/37458/commits/3f85c23f482de60e56292a91dadef1d3677665af.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: The phase2Snapshot  will be used by an async block - we can't close it right away. However, we close all resources when the finalize step completes which can be considered as a finally block.

**Comment by dnhatn** [comment]
Thanks @s1monw.

**Pull request merged**